tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/LetInTypeSynchronizer.scala

```diff
@@ -0,0 +1,75 @@
+package at.forsyte.apalache.tla.pp
+
+import at.forsyte.apalache.tla.lir.TypedPredefs.TypeTagAsTlaType1
+import at.forsyte.apalache.tla.lir.{LetInEx, OperEx, TlaDecl, TlaEx, TlaOperDecl, Typed, TypingException}
+import at.forsyte.apalache.tla.lir.transformations.{TlaExTransformation, TransformationTracker}
+import at.forsyte.apalache.tla.typecheck.etc.{Substitution, TypeUnifier}
+
+/**
+ * LetInTypeSynchronizer re-computes the types of any residual LET-IN defined operators, that were defined
+ * in the scope of a polymorphic operator (e.g. in FoldX). After inlining, any leftover LET-IN operators should have monotypes
+ */
+class LetInTypeSynchronizer(tracker: TransformationTracker) extends TlaExTransformation {
```

Comment:
This name is not clear to me. This transformation applies to a very specific pattern that comes from lambdas, or inlining of operators in higher-order operators. Shall we call it `LambdaTypeInliner`? It propagates the type from the body to the let-in definition.

Reply:
Sure